lib/graphql_metrics/tracer.rb

```diff
@@ -0,0 +1,127 @@
+# frozen_string_literal: true
+
+module GraphQLMetrics
+  class Tracer
+    # NOTE: Used to store timings from lexing, parsing, validation, before we have a context to store
+    # values in. Uses thread-safe Concurrent::ThreadLocalVar to store a set of values per thread.
+    cattr_accessor :pre_context
+
+    # NOTE: These constants come from the graphql ruby gem.
+    GRAPHQL_GEM_LEXING_KEY = 'lex'
+    GRAPHQL_GEM_PARSING_KEY = 'parse'
+    GRAPHQL_GEM_VALIDATION_KEYS = ['validate', 'analyze_query', 'analyze_multiplex']
+    GRAPHQL_GEM_TRACING_FIELD_KEYS = [
+      GRAPHQL_GEM_TRACING_FIELD_KEY = 'execute_field',
+      GRAPHQL_GEM_TRACING_LAZY_FIELD_KEY = 'execute_field_lazy'
+    ]
+
+    # TODO: can we implement this as an instance method? pass options instead of skip in context?
+    def self.trace(key, data)
+      # NOTE: Context doesn't exist yet during lexing, parsing.
+      possible_context = data[:query]&.context
+
+      skip_tracing = possible_context&.fetch(GraphQLMetrics::SKIP_GRAPHQL_METRICS_ANALYSIS, false)
+      return yield if skip_tracing
+
+      # NOTE: Not all tracing events are handled here, but those that are are handled in this case statement in
+      # chronological order.
+      case key
+      when GRAPHQL_GEM_LEXING_KEY
+        return setup_tracing_before_lexing { yield }
```

Comment:
I wonder if it is more efficient to materialize the block with `&block` and just pass it along, or if it's more efficient to do this and instantiate a second block just to yield to the first.

Reply:
I had block references, passed them as variables and used `.call` in my earlier iteration. @swalkinshaw suggested this approach and hinted it may be faster.

Digging into it a bit: even after some recent optimizations to passing `&block` and calling `.call`, `yield`ing directly still appears faster: https://github.com/JuanitoFatas/fast-ruby#proc--block

> or if it's more efficient to do this and instantiate a second block just to yield to the first.

I'm not sure I understand this suggestion. Can you provide a snippet?

Reply:
Before you were doing `resolve_block` without `&`. My original suggestion was to pass them as a block reference `&resolve_block` and `yield` still works and there's no need to explicitly `call` them

Reply:
Ah OK. Somehow I'd missed that.

I'm open to suggestions here. I suppose just `yield`ing everywhere without any block references is a bit more cryptic to readers?